c10/core/Device.h

```diff
@@ -86,6 +86,8 @@ struct C10_API Device final {
     return type_ == DeviceType::CPU;
   }
 
+  std::string str() const;
```

Comment:
Adding a function (which is so similar to `Device`'s `operator<<`) to something as central as `Device` for this case seems like overkill, I don't know if it's something we really need unless serialization time is a real bottleneck somewhere

Reply:
Thanks for the feedback.

Just checking: is the push-back related to an additional function in Device, or the duplicated code? The duplicated code aspect can be addressed by implementing operator<< in terms of str(), which I can do if that's the main objection.

fwiw, the benchmark I was looking at was torch::save() on a single 64-float tensor (as we might expect in a forward pass on a PS in DPER/etc), which is slower than it should be - 13usec, or about 200nsec amortized per float (most of it is fixed cost, the marginal cost/float beyond that is roughly 50nanos/float).

Given the performance goals, I was hoping to cull the easy wins like this from torch::save() - generally std::stringstream has a high constructor cost (thanks to dynamic_cast<>/etc internally), showing up in the pprof profiles. Given how widespread and convenient std::stringstream is, it doesn't make sure to eliminate altogether, but probably shouldn't be in any of the "hot" paths.

I am starting to wonder if it might make sense for the ThriftRpcAgent/et.al. to bypass torch::save()/load() for some specific cases (e.g. float contiguous cpu-device tensors), given the python-specific overhead, but mostly trying to figure out if there are good ways to squeeze costs in the current torch::save/load codepaths first. Anyway...

If you don't think this is architecturally appropriate, it's probably ok, though if there's an alternate way of expressing this logic that would be reasonable, I'm happy to change it! That said, we're running out of the super low hanging fruit for improving serialization speed, with keeping the existing format (though that said, some of the serialization benchmarks are 5x faster than a month ago, if you include FB-only crc diff).

Reply:
I think it is fine to just have operator<< for device use `str()`. That way we do not have divergence. Just adding `str` seems ok to me. It is disappointing that the idiomatic c++ string formatting is so slow.